cmd/daemonset-check/main.go

```diff
@@ -21,1118 +6,217 @@ import (
 
-	return allDS, nil
-}
-
-// Run implements the entrypoint for check execution
-func (dsc *Checker) Run(client *kubernetes.Clientset) error {
+	// Start listening to interrupts.
+	signalChan := make(chan os.Signal, 5)
+	go listenForInterrupts(signalChan)
 
-	// make a context for this run
-	dsc.ctx, dsc.cancelFunc = context.WithCancel(context.Background())
-
-	doneChan := make(chan error)
-
-	dsc.client = client
-
-	// run the check in a goroutine and notify the doneChan when completed
-	go func(doneChan chan error) {
-		err := dsc.doChecks()
-		doneChan <- err
-	}(doneChan)
+	// run check in background and wait for completion
+	runCheckDoneChan := make(chan error, 1)
+	go func() {
+		// Run daemonset check and report errors
+		runCheckDoneChan <- runCheck(ctx)
+	}()
 
-	var err error
-	// wait for either a timeout or job completion
+	// watch for either the check to complete or the OS to get a shutdown signal
 	select {
-	case <-time.After(dsc.Interval()):
-		// The check has timed out because its time to run again
-		dsc.cancelFunc() // cancel context
-		errorMessage := "Failed to complete checks for " + dsc.Name() + " in time!  Next run came up but check was still running."
-		//log.Errorln(dsc.Name(), errorMessage)
-		err = checkclient.ReportFailure([]string{errorMessage})
-		if err != nil {
-			log.Println("Error reporting failure to Kuberhealthy servers:", err)
-			return err
-		}
-		log.Println("Successfully reported failure to Kuberhealthy servers")
-	case <-time.After(dsc.Timeout()):
-		// The check has timed out after its specified timeout period
-		dsc.cancelFunc() // cancel context
-		errorMessage := "Failed to complete checks for " + dsc.Name() + " in time!  Timeout was reached."
-		//log.Errorln(dsc.Name(), errorMessage)
-		err = checkclient.ReportFailure([]string{errorMessage})
-		if err != nil {
-			log.Println("Error reporting failure to Kuberhealthy servers:", err)
-			return err
-		}
-		log.Println("Successfully reported failure to Kuberhealthy servers")
-	case err := <-doneChan:
-		dsc.cancelFunc()
-		if err != nil {
-			log.Println("Check failed with error:", err)
-			err = checkclient.ReportFailure([]string{err.Error()})
-			if err != nil {
-				log.Println("Failed to report error to Kuberhealthy servers:", err)
-				return err
-			}
-			log.Println("Successfully reported error to Kuberhealthy servers")
-			return nil
-		}
-		err = checkclient.ReportSuccess()
+	case err = <-runCheckDoneChan:
 		if err != nil {
-			log.Println("Error reporting success to Kuberhealthy servers:", err)
-			return err
+			reportErrorsToKuberhealthy([]string{"kuberhealthy/daemonset: " + err.Error()})
+		} else {
+			reportOKToKuberhealthy()
 		}
-		log.Println("Successfully reported success to Kuberhealthy servers")
+		log.Infoln("Done running daemonset check")
+	case <-signalChan:
+		log.Infoln("Received shutdown signal. Canceling context and proceeding directly to cleanup.")
+		reportOKToKuberhealthy()
```

Comment:
added a `reportOKToKuberhealthy` since if no report is made, kuberhealthy defaults the error to: `Check execution error: kuberhealthy/daemonset: timed out waiting for checker pod to report in` . not sure if we want to report an error instead